### ISSUE TITLE
Recognize functions like exit() as not returning

### ISSUE THREAD

**Issue by JukkaL** [python/mypy#2280]
This generates a warning when using `--warn-no-return`, even though it's safe:

```
def f(x: int) -> int:
    if x == 0:
        return x
    exit(1)
```

It would be nice if mypy knew about at least the most common stdlib functions that never return such as `sys.exit`.

**Comment by elazarg** [comment]
Something like this?

```Python
NoReturn = Union[()]
def exit(n: int) -> NoReturn: ...
```

**Comment by JukkaL** [comment]
That is one way to do it, but it may be a little confusing.

**Comment by JukkaL** [comment]
Having a general mechanism would be more flexible, but that's harder since we'll need to standardize on a solution (this has been discussed before and needs a decision here: https://github.com/python/typing/issues/165). Special casing a few functions like `sys.exit` would be a reasonable incremental improvement, at least for now.

**Comment by ddfisher** [comment]
Let's start by special casing the exit functions in the standard library.  That might be good enough for a while.

**Comment by gvanrossum** [comment]
Here's a strategy that could work. Make mypy understand Union[()] as the
magic, and update typeshed to mark up sys.exit(), os.exit() and
builtins.exit() with `-> Union[()]`. We won't need changes to typing.py
immediately (because this is only in the stubs) but we should still go
ahead and update typing.py to allow Union[()] -- currently it doesn't allow
this. We should also update PEP 484 to explain this.

Since it's relatively uncommon apart from those few builtins maybe we won't
need to pick a name for this -- Union[()] can be the idiomatic way to spell
it.

**Comment by gvanrossum** [comment]
We're going to spell this as `NoReturn`, initially in mypy_extensions. See https://github.com/python/typing/issues/165#issuecomment-271354561 and https://github.com/python/mypy/pull/2637.

**Comment by ilevkivskyi** [comment]
It looks like this issue is fixed now. At least the original example works correctly.

**Comment by ddfisher** [comment]
Yeah, this is fixed as of #2798.
